### ISSUE TITLE
Wrapper profile execution time is disproportionately affected by number of controls in the included profile

### ISSUE THREAD

**Issue by skpaterson** [inspec/inspec#3433]
### 🎛 Description

🐛 bug report

I have a wrapper profile that uses `require_controls` to execute only a subset of controls from an included profile containing 65.  The execution time for the subset seems to be adversely affected by the number of controls in the included profile.  This is especially true for cloud profiles, I suspect, because external calls are typically made when instantiating resources.  Taking a typical control that individually lasts a few seconds, overall execution time rises to several minutes for the wrapper profile.   Some degree of overhead is likely to be present for any profiles executed in this way.

### 🌍 InSpec and Platform Version

```
$ inspec --version
2.2.70
```

### 🤔 Replication Case

This “Replication case” is for the [GCP CIS Benchmark profile](https://github.com/chef-partners/cis-gcp-benchmark).  Each control here is a separate named file with a total of 65 in the repository.  Happy to provide access if required.

The wrapper profile is set up to point to a local directory, e.g. `inspec.yml` excerpt:
```
depends:
  - name: cis-gcp-benchmark
    path: /gcp-cis-benchmark/controls # let’s call this PARENT_CONTROLS_PATH
```

In my wrapper profile directory (say $WRAPPER_DIR) I request only two controls to be executed from the included profile e.g.

```
require_controls 'cis-gcp-benchmark' do
  control 'cis-gcp-benchmark-iam-1.1'
  control 'cis-gcp-benchmark-iam-1.2'
end
```

Running InSpec with only the two relevant controls inside $INCLUDED_CONTROLS_PATH works fine e.g.

```
$ ls $INCLUDED_CONTROLS_PATH/
iam_1_1.rb	iam_1_2.rb
```

Execution time is 7-8 seconds (output edited):

```
$ time bundle exec inspec exec . --attrs attributes.yaml -t gcp://

Profile: InSpec GCP CIS Environments Tests (cis-gcp-environments)
Version: 0.1.0
Target:  gcp://<service account>

     No tests executed.

Profile: InSpec GCP CIS Profile (cis-gcp-benchmark)
Version: 0.1.0
Target:  gcp://<service account>

  ✔  cis-gcp-benchmark-iam-1.1: Use corporate login credentials instead of Gmail accounts.
	…
  ×  cis-gcp-benchmark-iam-1.2: Ensure that multi-factor authentication is enabled for all non-service accounts (1 failed)
     ✔  User 12245 should have mfa enabled
     ×  User 112445 should have mfa enabled

Profile: Google Cloud Platform Resource Pack (inspec-gcp)
Version: 0.5.0
Target:  gcp://<service account>

     No tests executed.

Profile Summary: 1 successful control, 1 control failure, 0 controls skipped
Test Summary: 9 successful, 1 failure, 0 skipped

real	0m7.838s
user	0m2.296s
sys	0m0.362s
```

Now copy roughly half of the controls from the included profile into $INCLUDED_CONTROLS_PATH e.g.
```
$ ls $INCLUDED_CONTROLS_PATH/ | wc -l
      31
```

Re-running exactly the same command as before, the execution time jumps to almost 4 minutes:

```
real	3m54.761s
user	0m5.686s
sys	0m1.676s
```

Note this is with the wrapper profile in $WRAPPER_DIR unchanged. 

### 💁 Workaround

Similarly to the above, a wrapper profile can be used with a local directory containing only the relevant controls e.g. `inspec.yml` excerpt:

```
depends:
  - name: cis-gcp-benchmark
    path: /gcp-cis-benchmark/controls
```

Alternatively a custom profile with only the desired controls from the included profile could be created.  This creates duplication of controls but might be preferable in practice.

Aha! Link: https://chef.aha.io/features/SH-2426

**Comment by jquick** [comment]
This is currently a issue with require_controls/include_controls. We actually run them all we just only report on the ones listed. This is on our list to address post 3.0

**Comment by chris-rock** [comment]
Without digging in too much, I think the problem is only happening when profile dependencies are used in combination with filter table. Yes, resource instances are created but resources should only do heavy computation once once a method is called.

The current mechanism is not working if you do something like:
```
google_compute_zones(project: gcp_project_id).zone_names.each do |zone_name|
    ...
end
```

**Comment by jerryaldrichiii** [comment]
I agree with @chris-rock. This might be `filter_table` only or any code in the profile body (because it is executed during load). It looks like `--controls` works as expected when using a basic use case. See: https://github.com/inspec/inspec/issues/3091#issuecomment-443326217

**Comment by rmoles** [comment]
This issue is still present when making use of `include_controls` or `skip_controls`. Stumbled upon this issue when attempting to verify remediation using the CIS rhel7 profile. I attempted to wrap the profile to only run one control... eg.
```
  included_controls = %w[
    5.4.4_Ensure_default_user_umask_is_027_or_more_restrictive
  ]

  include_controls 'rhel7-cis-src' do
    included_controls.each { |ctrl| include_controls "xccdf_org.cisecurity.benchmarks_rule_#{ctrl}" }
  end
```

The issue also appears to be present when making use of `require_controls`

**Comment by clintoncwolfe** [comment]
To reiterate what Jared mentioned a while back, we suspect InSpec executes all controls in a wrapped profile, and simply filters the result. If you wrap a profile with 1000 controls and include 1, it will run 1000 controls and report on 1. This is fundamental to the way InSpec operates; this may or may not be intended - as there may be a delayed block execution effect that is intended but failing to occur.

The DSL names "include control" etc are probably misleading and should either be renamed or re-implemented, but re-implementation would also remove the ability to use arbitrary Ruby.

**Comment by ghost** [comment]
Any update on this? To use @clintoncwolfe 's example above - it would be nice to run only the controls specified instead of all & filter output.

**Comment by james-stocks** [comment]
I was discussing this with @skpaterson yesterday - the most common pitfall when trying to skip/exclude controls is that InSpec has to evaluate the source code you give it.

Frequently profile authors will mix Ruby statements in with InSpec's basic DSL - most commonly by including an `each` loop to define a number of `describe` tests; or by declaring a variable.

Currently *all* code must be contained within a describe block or inside a custom resource to make sure it gets deferred to after InSpec's initial load of the source code.

We definitely want to improve on this.

**Comment by Schwad** [comment]
Hello,

Although this is not perfect I had been looking into this code execution not long ago. You all probably know this but starting with this test example:

```ruby
title '/ profile'

control 'tmp-1.0' do
  impact 0.7
  title 'Create / directory'
  puts 'I am inside the control'
    describe file('/') do
      puts 'I am in the describe block'
      it { should be_directory }
    end
  end
```

Will normally return 

```
I am in the control block
I am in the describe block
```

When waived, the control only returns

```
I am in the control block
```

As stocksy said above, `describe` blocks get deferred when waived.

What if you spin up a dummy larger describe block for any code you don't want to be evaluated on a waiver? However, I am unsure if this is an accepted InSpec practice. (**It may well not be**: https://github.com/inspec/inspec/issues/630, although some old profiles seem to have done so.) Also, I am aware that it's not a comprehensive solution and is an additional step. :|

For some reason it did compile in this case.

I tested with this example below:

```ruby
title '/ profile'

control 'tmp-1.0' do
  impact 0.7
  title 'Create / directory'
  puts 'I am inside the control'
  describe 'do-not-execute-if-waived' do
    (1..1000).to_a.each do
      describe file('/') do
        sleep 1
        puts 'I am in the describe block'
        it { should be_directory }
      end
    end
  end
end

$ time inspec exec my_profile --waiver_file waivers.yml

#=> 1.73s user 0.67s system 96% cpu 2.481 total
```

In this contrived example, our code should have taken ~16 minutes to run (1000 describes with a 1 second pause) but it only took 2.481 seconds... As the waiver does ignore describe block contents.

Yes, it is an additional step, but I believe it would work today (until we can dig into this) for painful code chunks that are hitting run time. Stocksy may have thoughts on if this is a completely terrible idea 😅

**Comment by kekaichinose** [comment]
@Schwad @james-stocks I'm a little confused here... waivers not withstanding, I'm thinking this is a bug as I would expect `require_controls` (and consequently `include_controls` and `skip_controls`) to actually do what is specified. If the logic we've implemented with waivers is something that could be reused for say, `skip_controls`, then that might make sense. I def understand why there's confusion for users here.

**Comment by james-stocks** [comment]
The most common pitfall here is that users put Ruby in their InSpec control source code. By the nature of how InSpec works, that Ruby code will be evaluated at the same time that InSpec is evaluating what controls and tests are defined in the source code. 
There's nothing we can do in this case, we need users to back out any Ruby code in their InSpec controls and put them into custom resources.

One case that is a bug and we need to fix is with FilterTables - they load all data from the host or API being tested before waivers/skips/inheritance is evaluated - I opened https://github.com/inspec/inspec/issues/5133 for this specific case.

Another case that needs fixed is that InSpec resources must not load data from the host/API being tested in `initialize`. Again, InSpec can do nothing except execute this code immediately. https://github.com/inspec/inspec/issues/5134
